### PR TITLE
MySQL: Use BOOLEAN with comment hint to store boolean (issue #1205)

### PR DESCRIPTION
**Pull request by homersimpsons**
|      Q       |   A
|------------- | -----------
| Type         | feature/improvement
| BC Break     | yes
| Fixed issues | #1205 
#### Summary

Originally this PR should make use of `BIT(1)` to store boolean (instead of `TINYINT(1)`).

But as the real problem was the restriction about using TINYINT datatype I switched to a comment hint instead.

Reasons for this PR are explained in the issue #1205 (sum up here):
1. Programmer can't use TINYINT that can store 256 values, that dumbed down the field  
2. Programmer have to use SMALLINT which is twice larger, while TINYINT is much more adapted for a lot of case
3. Requiring SMALLINT doubles the size required and the input lag on selects

Furthermore this can cause issue on relation with TINYINT id column

**Note**: In MySQL `BOOLEAN` type internally map to `TINYINT(1)` and then the column is of type `TINYINT(1)`. An issue is already open in MySQL https://bugs.mysql.com/bug.php?id=8485

### CODE REVIEW

File: lib/Doctrine/DBAL/Platforms/MySqlPlatform.php [hunk 1]
```diff
@@ -1049,7 +1053,8 @@ public function getReadLockSQL()
     protected function initializeDoctrineTypeMappings()
     {
         $this->doctrineTypeMapping = [
-            'tinyint'       => 'boolean',
+            'bit'           => 'boolean',
+            'tinyint'       => 'tinyint',
```
Comment (morozov) [inline lib/Doctrine/DBAL/Platforms/MySqlPlatform.php]: Please provide some info on how it's expected to work with an already deployed schema.

File: lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php [hunk 2]
```diff
@@ -159,7 +162,10 @@ protected function _getPortableTableColumnDefinition($tableColumn)
                 break;
         }
 
-        if ($this->_platform instanceof MariaDb1027Platform) {
+        if ($type === Types::BOOLEAN) {
+            // As BIT (0b'0' or 0b'1') is used to store boolean values we must cast this default to a php boolean
+            $columnDefault = (bool) bindec(substr($tableColumn['default'], 2, -1));
```
Comment (morozov) [inline lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php]: A `BIT` column can be nullable. Looks like it's not taken into account.

File: lib/Doctrine/DBAL/Platforms/MySqlPlatform.php (hunk 1 above)
Comment (homersimpsons) [reply lib/Doctrine/DBAL/Platforms/MySqlPlatform.php]: I updated the PR to keep BOOLEAN (so TINYINT) and to use comment hint instead. On an already deployed schema, reverse engineering will break strict comparison (`=== true`).

File: lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php (hunk 2 above)
Comment (homersimpsons) [reply lib/Doctrine/DBAL/Schema/MySqlSchemaManager.php]: I updated the test to test for `null`. I also removed this part as it's no more relevant.

**Comment by homersimpsons** [comment]
I Updated this PR to use BOOLEAN datatype and Comment Hint the column. I also updated the description.

**Pull request closed without merging**